### DISCUSSION TITLE
Allow for multiple input_data in data_description.

### DISCUSSION THREAD

**Discussion by seanmcculloch** [AllenNeuralDynamics/aind-data-schema#1042] [Ideas]
**Is your feature request related to a problem? Please describe.**

When a derived data asset is a result of processing more than one data assets, there needs to be a strategy for documenting all relevant input data. An example of this is mfish round-to-round or other multi data-asset registration tasks.

permalink to relevant line: https://github.com/AllenNeuralDynamics/aind-data-schema/blob/687b7a3c4aa55552c89a80a3e6634a6c84f022fa/src/aind_data_schema/core/data_description.py#L153C5-L153C20

**Describe the solution you'd like**
The solution to this should take search-ability into account, along with compatibility to existing records and schema versions. I can see at least 3 potential solutions - alternate solutions are described below.

**Proposed Solution 1**. Add new field of _data_description.additional_input_data_ _-> Optional[List[str]]_.  When there are more than one input-data, the most important input data asset could be documented in _input_data_name_ (Chosen contextually per type of data asset being described). The rest of the input data could be put into _data_description.additional_input_data_.

By using _input_data_asset_ along with an optional field for more than one input-data is used, the _input_data_asset_ field could be used more consistently for searching by users. The added complexity for end-user queries would only apply to users who are expecting data assets with more than one input data, and should be familiar with the conventions of how _input_data_name_ and _additional_input_data_name_ are used. A naive search of _input_data_asset_ should still return closely related derived data with this strategy.

**Example: mFISH round-to-round registration results.**
A newly created derived data asset contains the displacement field transform and other info that maps data asset [HCR_BL6-000_2023-06-07_00-00-00/](http://aind-open-data.s3-us-west-2.amazonaws.com/HCR_BL6-000_2023-06-07_00-00-00/) into the coordinates of data asset [HCR_BL6-000_2023-06-1_00-00-00/](http://aind-open-data.s3-us-west-2.amazonaws.com/HCR_BL6-000_2023-06-1_00-00-00/)

Following Solution 1 we would notate:
input_data_name: "HCR_BL6-000_2023-06-07_00-00-00/"
additional_input_data_name: ["HCR_BL6-000_2023-06-01_00-00-00/"]

Where the convention for this type of data asset would be to put the moving-image data into **input_data_name** since the transform is intended to be applied to that data ("HCR_BL6-000_2023-06-07_00-00-00/") but the other dataset was used in it's calculation and must be documented as an input data in some form.

**Describe alternatives you've considered**

**Alternate Solution 2**. Change type of _data_description.input_data_name_ from _str_ to _List[str]_. The most straightforward change to the type of _input_data_name_ would change how user's interact with this field when searching, and would create inconsistencies with typing across versions.

**Alternate Solution 3.** Use _data_description.input_data_name_ as a comma-separated list within a single _str_. This would then require end-users to use regex-based queries to properly parse this field and search for data efficiently.

**Comment by tmchartrand** [comment]
As far as I can tell, _input_name_ as linked is not an actual schema field, but a property of the class that seems to be there just to help construct the name of derived data assets. The schema location for this would be in the _input_location_ field of the relevant DataProcess, which we currently have a PR up to allow lists in. We've also been discussing the option a composite input object with additional contextual metadata on why those inputs were selected: see #1167

**Comment by dbirman** [comment]
@seanmcculloch this is resolved in 2.0 btw, in `Code.input_data` you can now pass either a `DataAsset` or `CompositeData` which represent the two concepts you need, hopefully. I'll close this but feel free to re-open if you realize we're not supporting what you need.
